src/p11_child/p11_child_openssl.c

```diff
@@ -1745,7 +1745,7 @@ errno_t do_card(TALLOC_CTX *mem_ctx, struct p11_ctx *p11_ctx,
 
                 }
 
-                if ((info.flags & CKF_REMOVABLE_DEVICE) && (info.flags & CKF_TOKEN_PRESENT)) {
+                if ((info.flags & CKF_REMOVABLE_DEVICE)) {
```

Comment:
Hi,

I agree that the loop must be refactored to properly collect certificates from multiple sources however I wonder if instead of reverting the patch some extension like

````
if ((info.flags & CKF_REMOVABLE_DEVICE)) {
    removable_device_found = true;
    if (info.flags & CKF_TOKEN_PRESENT) {
        break;
    }
}
````
and then checking the boolean `removable_device_found` later on will fix the `--wait_for_card` use case as well?

bye,
Sumit

Reply:
Ok, I agree this can be made to work, but the change above would not be enough.

The current code for `wait_for_card` needs to know specifically which module contains the removable slot, so that it can wait for a token to be inserted.